src/ImageSharp/Processing/Processors/Filters/ReduceLuminanceProcessor.cs

```diff
@@ -0,0 +1,72 @@
+﻿// Copyright (c) Six Labors and contributors.
+// Licensed under the Apache License, Version 2.0.
+
+using System;
+using System.Numerics;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+using SixLabors.ImageSharp.Advanced;
+using SixLabors.ImageSharp.ParallelUtils;
+using SixLabors.ImageSharp.PixelFormats;
+using SixLabors.Primitives;
+
+namespace SixLabors.ImageSharp.Processing.Processors.Filters
+{
+    /// <summary>
+    /// Reduces the luminance of the input image, proportially to the luminance of each pixel.
+    /// </summary>
+    /// <typeparam name="TPixel">The pixel format.</typeparam>
+    internal class ReduceLuminanceProcessor<TPixel> : ImageProcessor<TPixel>
+        where TPixel : struct, IPixel<TPixel>
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="SixLabors.ImageSharp.Processing.Processors.Filters.ReduceLuminanceProcessor{TPixel}"/> class.
+        /// </summary>
+        /// <param name="amount">The strength of the luminance reduction filter.</param>
+        public ReduceLuminanceProcessor(float amount) => this.Amount = amount;
+
+        /// <summary>
+        /// Gets the strength factor for the luminance reduction processor.
+        /// </summary>
+        public float Amount { get; }
+
+        /// <inheritdoc/>
+        protected override void OnFrameApply(ImageFrame<TPixel> source, Rectangle sourceRectangle, Configuration configuration)
+        {
+            var interest = Rectangle.Intersect(sourceRectangle, source.Bounds());
+            int startX = interest.X;
+            float amount = this.Amount;
+            var factors = new Vector4(0.2125f, 0.7154f, 0.0721f, 0);
+
+            ParallelHelper.IterateRowsWithTempBuffer<Vector4>(
+                interest,
+                configuration,
+                (rows, vectorBuffer) =>
+                    {
+                        for (int y = rows.Min; y < rows.Max; y++)
+                        {
+                            Span<Vector4> vectorSpan = vectorBuffer.Span;
+                            int length = vectorSpan.Length;
+                            Span<TPixel> rowSpan = source.GetPixelRowSpan(y).Slice(startX, length);
+                            PixelOperations<TPixel>.Instance.ToVector4(configuration, rowSpan, vectorSpan);
+
+                            ref Vector4 baseRef = ref MemoryMarshal.GetReference(vectorSpan);
+                            Vector4Utils.Premultiply(vectorSpan);
+
+                            for (int i = 0; i < vectorSpan.Length; i++)
+                            {
+                                ref Vector4 v = ref Unsafe.Add(ref baseRef, i);
+                                float w = v.W;
+                                float luminance = 1 - (Vector4.Dot(v, factors) * amount);
```

Comment:
This isn't doing what the description in the linked document does.
https://microsoft.github.io/Win2D/html/T_Microsoft_Graphics_Canvas_Effects_LuminanceToAlphaEffect.htm

The linked effect changes the calculated alpha component to the calculated luminance and each pixel color component to 0.

Yours looks like you are multiplying the color components by the luminance and maintaining the original alpha value.

Reply:
Yeah, as I said, this is a luminance *reduction* filter, what you linked is just the luminance *to alpha* map. That's why I said the latter could be done with a single color matrix transform, but this effect can't, as it's a composition of multiple steps.
To reiterate, this effect does the equivalent of:
• Applying the luminance to alpha effect to get a luminance map
• Inverting the luminance map (to be used in the following step)
• Multiplying the resulting mapping over the original RGB channels, preserving the alpha

This effectively darkens just the brightest areas in the source images, leaving the rest intact (otherwise the whole image would just get darker, and the dark areas would just end up being black, if a simple brightness filter was used instead of this).

The effect you linked maps to the alpha channel just because that way you basically end up with a semi transparent luminance map for the input image, what I'm doing here instead is to also overlay that map on the original image, scaled by the input factor.

Reply:
Ah ok so it's a combination of the three methods. 

What I don't understand about your calculation then is that it the result doesn't spread evenly across all pixels like I would expect. (I can't find any papers or equivalent methods describing this)

The algorithm in my head would be the inverse of the grayscale matrix algorithm. Where 0 would be completely grayscale and 1 remained unchanged.

Reply:
Ok.... So even if we are simply adjusting the luminance this doesn't match any known formula I could fathom.

It seems to be nearly accurate if all the input vector values are the same but different should the individual component values change.

For example given the following input:
```
Input
<1, 1, 1, 1>
Amount
0.667
```
With your calculation the result looks like this.
```
Luminance Calc
<0.333, 0.333, 0.333, 1>
```

And the result of converting the Rgba input to YcbCr and adjusting the luminance is as follows

```
Round Trip Luminance Calc BTU.601
<0.3333333, 0.3333333, 0.3333333, 1>
Round Trip Luminance Calc BTU.709
<0.3333333, 0.3333333, 0.3333333, 1>
```

That looks good to me, the luminance change has been spread nicely across all pixel components.

However, given the following input:

```
Input
<0.25, 0.5, 0.75, 1>
Amount
0.667
```
Your calculation yields the following vector.
```
Luminance Calc
<0.1724779, 0.3449559, 0.5174338, 1>
```
Whereas round-tripping yields the following.
```
Round Trip Luminance Calc 601
<-0.0509804, 0.1960784, 0.4470589, 1>
```
```
Round Trip Luminance Calc 709
<-0.05882353, 0.1882353, 0.4392157, 1>
```

Here's the full dump from LinqPad. Maybe I'm doing something wrong but if so, I can't see it.

``` c#
void Main()
{
	// var factors = new Vector4(0.2125f, 0.7154f, 0.0721f, 0);

	// These factors are from Wikipedia.
	var factors = new Vector4(0.2126F, 0.7152F, 0.0722F, 0);
	"Input".Dump();
	var input =   new Vector4(.25F, .5F, .75F, 1);
	var v = input;
	v.Dump();
	float amount = .667F;
	"Amount".Dump();
	amount.Dump();
	float w = v.W;

	float luminance = 1 - (Vector4.Dot(v, factors) * amount);

	v *= luminance;
	v.W = w;

	"Luminance Calc".Dump();
	v.Dump();

	"Round Trip Luminance Calc 601".Dump();

	var ycbcr = RgbaToYCbCr(input, amount);
	YCbCrToRgba(ycbcr).Dump();

	"Round Trip Luminance Calc 709".Dump();
	var ycbcr709 = RgbaToYCbCr709(input, amount);
	YCbCrToRgba709(ycbcr709).Dump();
}

// Define other methods and classes here

// Jpeg YCbCr conversion algorithm 601.
public Vector4 RgbaToYCbCr(Vector4 input, float amount)
{
	// Invert.
	amount = 1 - amount;

	Vector4 rgb = input * 255F;
	float r = rgb.X;
	float g = rgb.Y;
	float b = rgb.Z;

	float y = (0.299F * r) + (0.587F * g) + (0.114F * b);
	float cb = 128F + ((-0.168736F * r) - (0.331264F * g) + (0.5F * b));
	float cr = 128F + ((0.5F * r) - (0.418688F * g) - (0.081312F * b));

	// Luminance Result inverted
	return new Vector4(y * amount, cb, cr, input.W);
}

// http://www.r-5.org/files/books/computers/algo-list/compression/Keith_Jack-Video_Demystified-EN.pdf
public Vector4 RgbaToYCbCr709(Vector4 input, float amount)
{
	// Invert.
	amount = 1 - amount;

	Vector4 rgb = input * 255F;
	float r = rgb.X;
	float g = rgb.Y;
	float b = rgb.Z;

	float y = (0.213F * r) + (0.715F * g) + (0.072F * b);
	float cb = 128F + ((-0.117F * r) - (0.394F * g) + (0.511F * b));
	float cr = 128F + ((0.511F * r) - (0.464F * g) - (0.047F * b));

	// Vector Result inverted
	return new Vector4(y * amount, cb, cr, input.W);
}

public Vector4 YCbCrToRgba(Vector4 input)
{
	float y = input.X;
	float cb = input.Y - 128F;
	float cr = input.Z - 128F;

	float r = (float)Math.Round(y + (1.402F * cr), MidpointRounding.AwayFromZero);
	float g = (float)Math.Round(y - (0.344136F * cb) - (0.714136F * cr), MidpointRounding.AwayFromZero);
	float b = (float)Math.Round(y + (1.772F * cb), MidpointRounding.AwayFromZero);

	return new Vector4(new Vector3(r, g, b) / 255F, input.W);
}

public Vector4 YCbCrToRgba709(Vector4 input)
{
	float y = input.X;
	float cb = input.Y - 128F;
	float cr = input.Z - 128F;

	float r = (float)Math.Round(y + (1.540F * cr), MidpointRounding.AwayFromZero);
	float g = (float)Math.Round(y - (0.459F * cr) - (0.183F * cb), MidpointRounding.AwayFromZero);
	float b = (float)Math.Round(y + (1.816F * cb), MidpointRounding.AwayFromZero);
	return new Vector4(new Vector3(r, g, b) / 255F, input.W);
}
```

Reply:
So, if you wanted to approximate this luminosity reduction effect with a color coordinate transformation, you should use the `RgbaToYCbCr709` conversion from your code (as that uses factors that are closer to the ones in the luminosity to alpha Win2D effect), plus you wouldn't need that `1 - amount` step in the conversion.

The reason for that is: in my case I'm doing the equivalent of mapping the luminance to a semi-transparent, dark texture that I then overlay on top of the original image (that's the reason for the `1 - factor` step, as I want the brightest pixels to become darker, while preserving the others)

If you remove that `1 - amount` step from your `RgbaToYCbCr709` method, you get:

```
Input
<0,25. 0,5. 0,75. 1>
Amount
0.667
Luminance Calc
<0,1724779. 0,3449558. 0,5174338. 1>
Round Trip Luminance Calc 709
<0,09411765. 0,345098. 0,5960785. 1>
```

A bit closer, but it still differs due to the fundamentally different processing method we're using here. I'm not sure though why you're trying to find another way to reproduce this exact effect, is it just out of curiosity? Because this is a specific pipeline I came up with that I found was very useful when, for instance, writing text over a background image, but I'm not sure whether there's an equivalent effect documented somewhere, as you said you couldn't find any papers or articles with these exact processing steps either.

If you want some exaples, here's a couple small screens from a section in the app I'm working on where I'm actually using this effect (this is from the Win2D/Composition pipeline, not with ImageSharp of course).
For each screen, the banner at the top is the unedited image, the one at the bottom is the one processed through the pipeline.

![image](https://user-images.githubusercontent.com/10199417/53496819-1b15bb00-3aa3-11e9-9a8f-5a95688e98bf.png)
![image](https://user-images.githubusercontent.com/10199417/53496825-1ea94200-3aa3-11e9-90f3-18063170eb36.png)

Notice how in the first screen, the text was very difficult to read due to the background image being very bright, and applying the filter resulted in a darker image, where the brightest areas were properly modified.
In the second screen instead, as the original image was pretty dark from the start, the effect left it almost untouched, without making it look almost entirely black, or too dark to be able to distinguish any details.

Reply:
The reason I’m trying to understand the algorithm and reproduce it is because its functionality does not conform to its name. What you are doing is some sort of histogram contraction not luminance reduction. 

When we add a processor to the core library it should contain functionality that is both correct and useful for all. Currently this feels like something that would be better suited to your personal project as it’s functionality is specific to your requirements only. 

Once we get the required bits of the library exposed you would be able to implement it as a processor with all the additional performance tools. Right now you should still be able to implement it but you would have to iterate per pixel and handle premultiplication yourself.

Reply:
Sure, I thought this might be useful for others as well since in general this is always handy whenever you have text over an image, but I get your point, that's fair.
I mean, I'd be all for just choosing a better name for it if that's an issue, but if you don't think this effect would be a good addition to the library per se, that's fine 😊

I guess you can close this PR if you want then, at least I'll be able to keep this code for the future so that when you eventually expose the processor and utilities APIs I'll be able to just reuse this processor with (hopefully) no additional work needed.